---
title: "Genotype"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
  encoding=encoding,
  output_file=file.path(dirname(input_file), out_dir, 'Genotype.html'))})
author: "Ying Wang"
date: "08/03/2021"
output:
  html_document:
   code_folding: hide
   toc: true
   toc_float: true
   number_sections: TRUE
---

```{r setup, include=FALSE}
WD="~/OneDrive - The University of Queensland/mater_projs/auti_proj/AutismSpeaks_AGP//"
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(warnings = FALSE)
knitr::opts_chunk$set(message  = FALSE)
#knitr::opts_chunk$set(eval = FALSE)
knitr::opts_knit$set(root.dir = WD)

library(reshape2)
library(ggplot2)
library(plyr)
library(knitr)
library(rmarkdown)
library(dplyr)
library(plotly)
library(gapminder)
library(DT)
require("ggrepel")
library(grid)
library(gridExtra)
library(png)
library(grid)
```

# Dataset Description

A total of **7880** individuals from **2611** families were genotyped on the Illumina Human1M-Duov3_B or the Human1Mv1_C.

 - 4901 males, 2979 females.
 - 2571 trios, 36 quads, 1 pentas, 3 hexs.
 - 947,233 SNPs were genotyped. 
 - Coordinates were based on Build36.
 
 
 
```{r dataset_check, eval=FALSE, include=FALSE}
setwd("/QRISdata/Q3457/AGP/GenotypeFiles/")
fam1 <- read.table("phg000294.v1.NIMH_AutismGenomeProject_v2.genotype-calls-matrixfmt.c1.ARD/SubjectGenotypeFile.AGP.phs000267.v1.p1.20121018.fam", stringsAsFactors = F) ##7880 individuals
fam2 <- read.table("phg000143.v2.NIMH_AutismGenomeProject.genotype-calls-matrixfmt.Human1Mv1_C.c1.DS-ASD/AGP.phs000267.v1.p1.20110621.clean_3891.fam", stringsAsFactors = F) ##3981 individuals

coms <- intersect(fam1$V2, fam2$V2) ##3981 individuals, fam2 were totally included in fam1
# 
# table(fam1$V5)
# 
#    1    2 
# 4901 2979

length(unique(fam1$V1)) ##2611 families
sums <- data.frame(t(table(fam1$V1)))
table(sums$Freq)
#    3    4    5    6 
# 2571   36    1    3 

```

 ***
 
<br>

# Raw Genotype QC 
## Sex Check  {.tabset .tabset-fade .tabset-pills}

- 141 PRROBLEM
  - 115 with complete missing chrX genotypes.
   - 26 with chrX-F ranging from 0.20 to 0.62
 
### Mismatch summary
```{r}
sex <- read.table("../outputs/SubjectGenotypeFile.AGP.phs000267.v1.p1.20120806_checksex_pruned.sexcheck", header = T, stringsAsFactors = F)
sex1 <- sex[sex$STATUS == "PROBLEM" & !is.na(sex$F),]

datatable(sex1, rownames = FALSE, filter="top", caption = htmltools::tags$caption(
        style = 'caption-side: top; text-align: center; color:black; 
        font-size:200% ;',"Individuals with discordant sex information"), options = list(pageLength = 5, scrollX=T) )
```


***
 
<br>
 
 
### ChrX F distributions
```{r sex-dist, out.width="100%", out.height="100%", warning=FALSE}
grid.raster(readPNG("figures/sexD_pruned.png"))
```

***
 
<br>
 

## Pariwise IBD estimation {.tabset .tabset-fade .tabset-pills}

- Relationships (RT): OT (Others), FS (Full Siblings), PO (Parent Offspring) 
- family ID 483 & 1012 has potential issue
  - FID:483 with IBD sharing = 1 between IID:328 (Female) and IID:1491 (Female), which were supposed to be ~0. Are they MZ? same individual? The genotype missing rates are 0.1185 and 0.1186 for IID:483_328 and IID:483_1491, respectively. [Drop IID:483_1491, IID:483_4371 and IID:483_993.]
  -  FID:1012 with IBD sharing = 0.57 between IID:2319 (Female) and IID: 3612 (Female). They are supposed to be FS but recruited into the same FID. Other kinship between FID:1012 can be confirmed. 
- IBS sharing for other pairs: ranging from 0.44 to 0.58 in FS, from 0.50 to 0.59 in PO, from 0 to 0.12 in OT (which indicating inbreeding between some parents.)


### Estimated pairwise IBD distributions
```{r IBD-dist, out.width="100%", out.height="100%", warning=FALSE}
grid.raster(readPNG("figures/IBD.png"))
```

***
 
<br>
 
 
### Family 483 & 1012
```{r read_genome}
genome <- read.table("../outputs/SubjectGenotypeFile.AGP.phs000267.v1.p1.20120806_pruned_IBD.genome", header = T, stringsAsFactors = F)
genome1 <- genome[genome$FID1 == 483 | genome$FID1 == 1012,]

datatable(genome1, rownames = FALSE, filter="top", caption = htmltools::tags$caption(
        style = 'caption-side: top; text-align: center; color:black; 
        font-size:200% ;'), options = list(pageLength = 5, scrollX=T) )
```

***
 
<br>
 

## Individual genome-wide heterozygosity {.tabset .tabset-fade .tabset-pills}

### Genome-wide heterozygosity VS missing rates 
```{r genomewideHet, out.width="100%", out.height="100%", warning=FALSE}
grid.raster(readPNG("figures/Het_VS_missing2.png"))
```

***
 
<br>
 
**Note** that samples were genotyped in the Human1M-Duov3_B or the Human1Mv1_C. Genotypes for these individuals are an union of the genotypes from both platforms. For missingness, only the intersecting SNPs between two arrays were used.

```{r include=F, eval=FALSE}
# - In the final merged dataset 947,233 markers are represented. Of these 938,130 are represented on the  Human1M-Duov3_B while 858,052 are on the Human1Mv1_C. 
# - For individuals on the Human1M-Duov3_B the missingness rate is at least 1% while for individuals genotyped on the Human1Mv1_C it is at least 9%. 
# - Individuals were checked to have no more than a 5% missingness rate for the platform they were genotyped on. 

```


***
 
<br>
 

### Genome-wide F VS missing rates 
```{r genomewideF, out.width="100%", out.height="100%", warning=FALSE}
grid.raster(readPNG("figures/inbreeding_VS_missing2.png"))
```

***
 
<br>
 


# Imputation

## Pre-imputation
The imputation pipeline follows that used for SSC dataset. A total of **7769** individuals and ~784K autosomal, ~22K chrX SNPs were used for further impution.

- filters: \--geno 0.05 \--mind 0.1 \--maf 0.01 \--hwe 1e-6
  - 111 people removed due to missing genotype data (--mind). Their missingness rates ranging from 0.7 to 1.
  - Total genotyping rate in remaining samples is 0.914029.
  - 124565 variants removed due to missing genotype data (--geno).
  - 15633 variants removed due to Hardy-Weinberg exact test.


***
 
<br>
 
## After Imputation
### Frequency distribution
- ~7.6M SNPs overlapped SNPs between AGP_imputed and HRC_WGS (passing filters: \--geno 0.05 \--maf 0.01 \--hwe 1e-6)
- based on same allele
- **0 SNPs with MAF difference > 0.2**

```{r frq_dis, out.width="100%", out.height="100%", warning=FALSE}
grid.raster(readPNG("figures/AGP_frq_aftImp.png"))
```

***

<br>

### PCA 

- Project the first 3 PCs based on pruned HapMap3 SNPs onto 1000G
- Using K-means to calculate distance
- Assign ancestry based on posterior probability 0.9
  - 6548 Europeans (EUR), 625 Americans (AMR),  123 South-Asians (SAS), 99 East-Asians (EAS) and 141 Africans (AFR).

```{r plot_anc_all, out.width="80%", out.height="60%", echo=FALSE}
toAssign <- read.table("../outputs/agp_anc.txt", header = T, stringsAsFactors = F)

ggplot(data = toAssign, aes(x = PC1, y = PC2, color = Assigned)) +
  geom_point() +
  ylab("PC2") +
  xlab("PC1")
```

***

<br>

